db/tidb/bank_service_quality.go

```diff
@@ -0,0 +1,206 @@
+package tidb
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"time"
+
+	"github.com/pingcap/tipocket/pkg/core"
+)
+
+type rtoMetrics struct {
+	nemesis          *core.NemesisGeneratorRecord
+	p80              time.Duration
+	p90              time.Duration
+	p99              time.Duration
+	totalMeasureTime time.Duration
+}
+
+type bankServiceQualityChecker struct {
+	profFile string
+}
+
+type countWindow struct {
+	series []time.Time
+	pos    int
+}
+
+func (f *countWindow) count(duration time.Duration) (*time.Time, int, error) {
+	if f.pos >= len(f.series) {
+		return &f.series[len(f.series)-1], 0, fmt.Errorf("window has ended")
+	}
+
+	var count = 0
+	var pos = f.pos
+	var startTime = f.series[pos]
+
+	defer func() {
+		f.pos++
+	}()
+
+	for pos < len(f.series) {
+		t := f.series[pos]
+		if t.After(startTime.Add(duration)) {
+			break
+		}
+		count++
+		pos++
+	}
+	return &startTime, count, nil
+}
+
+// Check checks the bank history and measure failure requests count before and after nemesis was injected.
+// When the avg failure count of one continuous minute is below 0.8 * the failure count before nemesis, we say the the system
+// recover to p80 line from that time point. And p90 or p99 are similar.
+func (c bankServiceQualityChecker) Check(_ core.Model, ops []core.Operation) (bool, error) {
+	var nemesisRecord core.NemesisGeneratorRecord
+	var duration = time.Minute
+	var failCount float64
+	var failCountPerMinuteBeforeInject float64
+
+	var injectionTime time.Time
+	//var recoveryTime time.Time
+
+	var p80 *time.Duration
+	var p90 *time.Duration
+	var p99 *time.Duration
+
+	startTime := ops[0].Time
+	idx := 0
+	for idx < len(ops) {
+		op := ops[idx]
+		if op.Action == core.ReturnOperation {
+			response := op.Data.(bankResponse)
+			if response.Unknown {
+				failCount++
+			}
+		} else if op.Action == core.InvokeNemesis {
+			injectionTime = op.Time
+			nemesisRecord = op.Data.(core.NemesisGeneratorRecord)
+			break
+		} else if op.Action == core.RecoverNemesis {
+			//recoveryTime = op.Time
+			//break
+		}
+		idx++
+	}
+	failCountPerMinuteBeforeInject = failCount / injectionTime.Sub(startTime).Minutes()
+	if idx == len(ops) {
+		log.Panic("expect more events after recovery from nemesis")
+	}
+
+	var window *countWindow
+	metrics := rtoMetrics{nemesis: &nemesisRecord}
+
+	window, metrics.totalMeasureTime = convertToFailSeries(ops[idx:])
+	if window != nil {
+		for true {
+			startTime, failCount, err := window.count(duration)
```

Comment:
making it an iterable object maybe more understandable.

Reply:
That's a good idea